src/python-common/ceph/deployment/service_spec.py

```diff
@@ -140,17 +140,123 @@ def validate(self) -> None:
         assert_valid_host(self.hostname)
 
 
+HostPatternType = Union[str, None, Dict[str, Union[str, bool, None]], "HostPattern"]
+
+
+class PatternType(enum.Enum):
+    fnmatch = 'fnmatch'
+    regex = 'regex'
+
+
+class HostPattern():
+    def __init__(self,
+                 pattern: Optional[str] = None,
+                 pattern_type: PatternType = PatternType.fnmatch) -> None:
+        self.pattern: Optional[str] = pattern
+        self.pattern_type: PatternType = pattern_type
+        self.compiled_regex = None
+        if self.pattern_type == PatternType.regex and self.pattern:
+            self.compiled_regex = re.compile(self.pattern)
+
+    def filter_hosts(self, hosts: List[str]) -> List[str]:
+        if not self.pattern:
+            return []
+        if not self.pattern_type or self.pattern_type == PatternType.fnmatch:
+            return fnmatch.filter(hosts, self.pattern)
+        elif self.pattern_type == PatternType.regex:
+            if not self.compiled_regex:
+                self.compiled_regex = re.compile(self.pattern)
+            return [h for h in hosts if re.match(self.compiled_regex, h)]
+        raise SpecValidationError(f'Got unexpected pattern_type: {self.pattern_type}')
+
+    @classmethod
+    def to_host_pattern(cls, arg: HostPatternType) -> "HostPattern":
+        if arg is None:
+            return cls()
+        elif isinstance(arg, str):
+            return cls(arg)
+        elif isinstance(arg, cls):
+            return arg
+        elif isinstance(arg, dict):
+            if 'pattern' not in arg:
+                raise SpecValidationError("Got dict for host pattern "
+                                          f"with no pattern field: {arg}")
+            pattern = arg['pattern']
+            if not pattern:
+                raise SpecValidationError("Got dict for host pattern"
+                                          f"with empty pattern: {arg}")
+            assert isinstance(pattern, str)
+            if 'pattern_type' in arg:
+                pattern_type = arg['pattern_type']
+                if not pattern_type:
+                    return cls(pattern, pattern_type=PatternType.fnmatch)
+                else:
+                    if pattern_type == 'fnmatch':
+                        return cls(pattern, pattern_type=PatternType.fnmatch)
+                    elif pattern_type == 'regex':
+                        return cls(pattern, pattern_type=PatternType.regex)
+                    else:
+                        raise SpecValidationError("Got dict for host pattern "
+                                                  f"with unknown pattern type: {arg}")
```

Comment:
nit: I think there's no need for the `else` in this case, this code is equivalent to:

```
if 'pattern_type' in arg:
    pattern_type = arg['pattern_type']

    if not pattern_type:
        return cls(pattern, pattern_type=PatternType.fnmatch)
    elif pattern_type == 'fnmatch':
        return cls(pattern, pattern_type=PatternType.fnmatch)
    elif pattern_type == 'regex':
        return cls(pattern, pattern_type=PatternType.regex)
    else:
        raise SpecValidationError(
            f"Got dict for host pattern with unknown pattern type: {arg}"
        )
```

Reply:
done, code is now structured like your comment